### PR TITLE
Landing Page Form Submission

### PR DESCRIPTION
**Pull request by McCambley**
# Overview

### Bullet Pointed Updates
_A more detailed breakdown of changes that made the form submission possible_
1) Create `PATCH` route and controller to handle updates  to investor `formInputs` 
   - Used in new user registration and future profile editing
2) Reconfigure authentication middleware to throw and Error instead of redirecting 
   - Had previously thrown a CORS error that was overlooked. Now the redirection is handled in the frontend.
3) Adjust routes to use centralized error handling with `next()` 
4) Add a `useEffect` to `App.js` to `PATCH` the loggedIn investors `formInputs` with an array previously stored in `localStorage` during form submission
   - The original hope was to send this data along earlier in the passport process but that proved dumb and impossible
5) Expand the Api on the frontend with `updateUser`
6) Expand the Api on the frontend with `validateUser`
7) Refactor `userContext` into an `App` level state value
   - Context is excessive for the needs of this project, and we need to be able to update the context in a way that is far easier to do if the user data is stored in the `App` level rather than a higher level `Context` 

### Flow of data
_High level view of what this PR actually does_

What we want is potential investors to use the form on the landing page to choose their desired form inputs, things that they actually care about seeing from startups that will fill out their form. What's cool about this app from a user perspective is that investors can create their form before signing up. What's not cool from a developer perspective is that investors can create their form before signing up... how do we assign a form to a user that doesn't exist yet. 
1) From the landing page, investors click the form fields they care about. A state variable stored these preferences in an array of strings.
2) Investors click "Finish creating the form" when they're satisfied and that array of strings is stored to localStorage
3) Investors go through the process of redirects that lets them signup with social accounts. When the login is successful, they're redirected back to the app.
4)  At that point, a `useEffect` sends that stored array of strings in a `PATCH` to update that investor object that was _just_ created. 
5) If the previous step is successful, the `user` variable (refactored to `App` level state variable instead of `Context`) is updated with the now updated user

[Loom Demo](https://www.loom.com/share/4e21d1e06d1f4990862a7dbe78545088)

# I follow best practices 

- [x] My code is covered by unit tests 
- [x] This PR is less than [300 lines of code](https://www.atlassian.com/blog/git/written-unwritten-guide-pull-requests)
- [x] My code [smells good](https://en.wikipedia.org/wiki/Code_smell) 
- [x] I have tested my PR locally and proven it works. 
- [x] There is [no copy-paste](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)

### CODE REVIEW

File: frontend/src/components/landing/Landing.js [hunk 1]
```diff
@@ -96,7 +96,7 @@ function Landing({ setLoggedIn }) {
       </FounderContainer>
       <Cards>
         <Card1 translate>
-          <CardHeading $width='316px'>1. You create a form</CardHeading>
+          <CardHeading $width="316px">1. You create a form</CardHeading>
```
Comment (rileydanejohnston) [inline frontend/src/components/landing/Landing.js:99]: Why double quotes > single quotes?

File: frontend/src/components/landing-form/LandingForm.js [hunk 2]
```diff
@@ -124,6 +125,15 @@ export default function LandingForm({ setLoggedIn }) {
     });
   };
 
+  // update investors desired formFields when button is selected
+  useEffect(() => {
```
Comment (rileydanejohnston) [inline frontend/src/components/landing-form/LandingForm.js:129]: Very cool

**Review by rileydanejohnston — CHANGES_REQUESTED** [review]
I think this looks good. I made a few comments, mostly just questions I had to help me understand what's going on.

File: frontend/src/components/landing/Landing.js (hunk 1 above)
Comment (McCambley) [reply frontend/src/components/landing/Landing.js:99]: To be honest I'm not sure, my prettier is set to auto correct to single quotes so I'm not sure why this happened...

File: frontend/src/components/app/App.js [hunk 3]
```diff
@@ -13,21 +13,55 @@ import NotFound from '../not-found/NotFound';
...
+      .catch((error) => {
+        setUser(null);
+        return history.push('/welcome');
+      });
+  }, []);
 
   React.useEffect(() => {
     // check for populated user object
-    if (!!user && !!user.email) {
+    if (!!user) {
```
[the remark this answers is not in the document]
Comment (McCambley) [reply frontend/src/components/app/App.js:41]: Double exclamation turns things to boolean. If `user.email === "email@thing.com"`, `!!user.email === true`. If `user.email === undefined`, `!!user.email === false`

File: backend/controllers/investors.js [hunk 4]
```diff
@@ -48,4 +49,34 @@ function createOrUpdateInvestor(profile, done) {
...
+      if (!investor) {
+        throw new ErrorHandler(404, 'User not found');
+      }
+      // update stored investor
+      req.session.passport.user = investor;
+      // send updated user information
+      res.send({ data: investor });
+    })
+    .catch((error) => {
+      if (error.name === 'CastError') {
```
[the remark this answers is not in the document]
Comment (McCambley) [reply backend/controllers/investors.js:74]: These get sent to our centralized error handler. `next(err)` is expresses built in way of passing errors down the process to one central place. Makes scaling up and error debugging easier. I think that's what you're asking?

File: backend/routes/investors.js [hunk 5]
```diff
@@ -1,9 +1,10 @@
 const express = require('express');
+const { updateInvestor } = require('../controllers/investors');
 
 const router = express.Router();
 
-// Verify the user from frontend/src/contexts/UserContext
-router.get('/', (req, res) => {
+// Verify the user from App
+router.get('/', (req, res, next) => {
```
[the remark this answers is not in the document]
Comment (McCambley) [reply backend/routes/investors.js:7]: Doesn't need to be, but I think it's commonplace to have it there. When centralized error handling is implemented, `next(err)` is the built in way to handle those errors. If we were to do some error handling on this route (I probably should), I'd send the error in `next()`

**Review by rileydanejohnston — APPROVED** [review]
Looks good!

**Pull request merged**